src/types/Utf16Parser.cpp

```diff
@@ -15,35 +16,41 @@
 // - A view into the string given of just the next codepoint unit.
 std::wstring_view Utf16Parser::ParseNext(std::wstring_view wstr)
 {
-    size_t pos = 0;
-    size_t length = 0;
-
-    for (auto wch : wstr)
+    for (size_t pos = 0; pos < wstr.size(); ++pos)
     {
+        const auto wch = wstr.at(pos);
+
+        // If it's a lead and followed directly by a trail, then return the pair.
+        // If it's not followed directly by the trail, go around again and seek forward.
         if (IsLeadingSurrogate(wch))
         {
-            length++;
+            // Try to find the next item... if it isn't there, we'll go around again.
+            const auto posNext = pos + 1;
+            if (posNext < wstr.size())
+            {
+                // If we found it and it's trailing, return the pair.
+                const auto wchNext = wstr.at(posNext);
+                if (IsTrailingSurrogate(wchNext))
+                {
+                    return wstr.substr(pos, 2);
+                }
+            }
+            // If we missed either if in any way, we'll fall through and go around again searching for more.
         }
+        // If it's just a trail at this point, go around again and seek forward.
         else if (IsTrailingSurrogate(wch))
         {
-            if (length != 0)
-            {
-                length++;
-                break;
-            }
-            else
-            {
-                pos++;
-            }
+            continue;

```

Comment:
Just so I understand (and it seems tests uphold this assertion):

```
TRAIL             -> REPLACEMENT
TRAIL TRAIL TRAIL -> REPLACEMENT
TRAIL U0041       -> U0041
```

Reply:
Yes, that's what I was going for... for now at least. We'll tweak it in later if my assertions prove incorrect in some other context.